File: src/org/dreamerslab/smtp/Smtp.java
package org.dreamerslab.smtp;

import android.app.Activity;
import com.google.appinventor.components.annotations.SimpleEvent;
import com.google.appinventor.components.annotations.SimpleFunction;
import com.google.appinventor.components.runtime.AndroidNonvisibleComponent;
import com.google.appinventor.components.runtime.ComponentContainer;
import com.google.appinventor.components.runtime.EventDispatcher;
import com.google.appinventor.components.runtime.util.AsynchUtil;

import javax.mail.*;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;
import java.util.Date;
import java.util.Properties;

public class Smtp extends AndroidNonvisibleComponent {

    private final Activity activity;

    public Smtp(ComponentContainer container) {
        super(container.$form());
        activity = container.$context();
    }

    @SimpleFunction(description = "Send message using smtp.")
    public void SendMessage(
            final String smtpHost,
            final String smtpPort,
            final String user,
            final String password,
            final String senderEmail,
            final String receiverEmail,
            final String emailCc,
            final String subject,
            final String content
    ) {
        final Properties properties = new Properties();
        properties.putIfAbsent("mail.smtp.host", smtpHost);
        properties.putIfAbsent("mail.smtp.port", smtpPort);
        properties.putIfAbsent("mail.smtp.auth", "true");
        properties.putIfAbsent("mail.smtp.starttls.enable", "true");

        final Session session = Session.getInstance(properties, new Authenticator() {
            @Override
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication(user, password);
            }
        });

        try {
            final MimeMessage mimeMessage = new MimeMessage(session);
            mimeMessage.setFrom(senderEmail);
            mimeMessage.addRecipients(Message.RecipientType.TO, InternetAddress.parse(receiverEmail));
            if (!emailCc.isEmpty()) {
                mimeMessage.addRecipients(Message.RecipientType.CC, InternetAddress.parse(emailCc));
            }

            mimeMessage.setSubject(subject);
            mimeMessage.setText(content);
            mimeMessage.setSentDate(new Date());

            final Runnable runnable = () -> {
                try {
                    final Transport transport = session.getTransport("smtp");
                    transport.connect();
                    transport.sendMessage(mimeMessage, mimeMessage.getAllRecipients());
                    transport.close();

                    activity.runOnUiThread(this::OnSuccess);
                } catch (MessagingException e) {
                    OnError(e.getMessage());
                }
            };

            AsynchUtil.runAsynchronously(runnable);
        } catch (MessagingException e) {
            OnError(e.toString());
        }
    }

    @SimpleEvent(description = "Event raised when message sending fails.")
    public final void OnError(String msg) {
        EventDispatcher.dispatchEvent(this, "OnError", msg);
    }

    @SimpleEvent(description = "Email sent successfully")
    public final void OnSuccess() {
        EventDispatcher.dispatchEvent(this, "OnSuccess");
    }
}

File: src/org/dreamerslab/smtp/Smtp.kt
package org.dreamerslab.smtp

import com.google.appinventor.components.annotations.SimpleEvent
import com.google.appinventor.components.annotations.SimpleFunction
import com.google.appinventor.components.runtime.AndroidNonvisibleComponent
import com.google.appinventor.components.runtime.ComponentContainer
import com.google.appinventor.components.runtime.EventDispatcher
import com.google.appinventor.components.runtime.util.AsynchUtil
import com.google.appinventor.components.runtime.util.YailList
import java.util.*
import javax.mail.*
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeMessage

@Suppress("FunctionName")
class Smtp(container: ComponentContainer) : AndroidNonvisibleComponent(container.`$form`()) {

    private val activity = container.`$context`()
    private var session: Session? = null

    @SimpleFunction(description = "Create session.")
    fun CreateSession(
        smtpHost: String,
        smtpPort: String,
        user: String?,
        password: String
    ) {
        val properties = Properties().apply {
            getOrPut("mail.smtp.host") { smtpHost }
            getOrPut("mail.smtp.port") { smtpPort }
            getOrPut("mail.smtp.auth") { "true" }
            getOrPut("mail.smtp.starttls.enable") { "true" }
        }
        val authenticator = object : Authenticator() {
            override fun getPasswordAuthentication(): PasswordAuthentication {
                return PasswordAuthentication(user, password)
            }
        }
        session = Session.getInstance(properties, authenticator)
    }

    @SimpleFunction(description = "Send message using smtp.")
    fun SendMessage(
        senderEmail: String,
        receiverEmails: YailList,
        emailCcs: YailList,
        subject: String,
        content: String
    ) {
        if (session == null) {
            OnError("Session not initialized.")
            return
        }
        try {
            val mimeMessage = MimeMessage(session).apply {
                setFrom(senderEmail)
                receiverEmails.toStringArray()
                    .map { InternetAddress(it) }
                    .forEach {
                        addRecipient(Message.RecipientType.TO, it)
                    }

                emailCcs.toStringArray()
                    .map { InternetAddress(it) }
                    .forEach {
                        addRecipient(Message.RecipientType.CC, it)
                    }

                this.subject = subject
                setText(content)
                sentDate = Date()
            }

            AsynchUtil.runAsynchronously {
                try {
                    session?.getTransport("smtp")?.apply {
                        connect()
                        sendMessage(mimeMessage, mimeMessage.allRecipients)
                        close()
                    }
                    activity.runOnUiThread { OnSuccess() }
                } catch (e: MessagingException) {
                    OnError(e.message)
                }
            }
        } catch (e: MessagingException) {
            OnError(e.toString())
        }
    }

    @SimpleEvent(description = "Event raised when message sending fails.")
    fun OnError(msg: String?) {
        EventDispatcher.dispatchEvent(this, "OnError", msg)
    }

    @SimpleEvent(description = "Email sent successfully")
    fun OnSuccess() {
        EventDispatcher.dispatchEvent(this, "OnSuccess")
    }
}